### ISSUE TITLE
RNN or so to predict mouse motion

### ISSUE THREAD

**Issue by MariasStory** [BrainJS/brain.js#61]
Hi,

Not an issue per say.
It would be very nice to have a simple RNN or so to predict mouse motion.
Did someone see/did such example in a browser?
I have seen such example in c (don't have the reference now).

I don't look for a fastest or smartest example. Just one that work ;-)

Thanks.

**Comment by robertleeplummerjr** [comment]
What would input data look like?  If we were to go super specific like pixel position, training would be something like:
```js
var net = new brain.NeuralNetwork();
net.train([
  {
    input: [whereTheyWhereX / thingWidth, whereTheyWhereY / thingHeight],
    output: [whereTheyAreX / thingWidth, whereTheyAreY / thingHeight]
  }
]);
```

So setup a environment to capture mouse moves, use the onmousemove event, and the more input data you give it, the more precise it _should_ be.

**Comment by robertleeplummerjr** [comment]
If you were to use an RNN, such as a LSTM, my gut reaction is to do something more like:
```js
var net = new brain.recurrent.RNN();
net.train([
  {
    input: ['slowPositiveX', 'neutralY', 'fastPositiveX', 'negativeSlowY', 'neutralX', 'fastNegativeY' ...], //keep going until mouse rests 
    output: ['slowPositiveX', 'neutralY']
  }
]);
```
Then iterate through the list of positions and normalize the values to one of the enums above or any you come up with.  This extracts the math, so the neural net can work out the principle rather than both math and principle.  Then when you go to predict, you'd need current location, and last location, normalized, then on data output, you can denormalize them to the location of the current mouse position.

huzzah!

**Comment by robertleeplummerjr** [comment]
I forgot to mention, then you could do:
```js
var net = new brain.recurrent.RNN();
var positionX = relativePosition(previousX, thisX);
var positionY = relativePosition(previousY, thisY, true);
var output = net.run([positionX, positionY]);
```

`relativePosition` would simply accept two numbers, and output a string, such as `'slowPositiveX'` or `'slowPositiveY'`.
`output` here would simply be a key.

**Comment by robertleeplummerjr** [comment]
So, as it turns out, I couldn't resist and ended up writing a solution for mouse prediction.  Let me know what you think: https://jsfiddle.net/robertleeplummerjr/wvkbojr7/

You paint in red, and after the net is trained, it paints along with you in green.

If time equals money...
thinking about it = one week
writing it = 30 minutes
making training data = a minute
training it = 2 seconds
creating mathematical art = priceless

This was fun!

**Comment by robertleeplummerjr** [comment]
![predict](https://cloud.githubusercontent.com/assets/679099/23110544/30d58072-f6f0-11e6-9f30-24035daa2c9d.png): What it output for me.

**Comment by robertleeplummerjr** [comment]
Should be a little more accurate: https://jsfiddle.net/robertleeplummerjr/wvkbojr7/1/

**Comment by MariasStory** [comment]
Hi @robertleeplummerjr ,

This looks interesting.
I did something similar/different:
1) Record last N(~20) points of your mouse position.
2) Train a network with last T(~10) points and use current mouse position as expected result.
3) Activate network with last T points of your mouse position.
I do it for every mouse move and was able to use quite a complex model (on my system I5).
I expected to get a prediction of your mouse position ~10 points ahead and a continuously improving network. In reality I see something less exiting:
1) At the beginning the network performs almost as yours (somewhat better). The predictions leg mouse motion.
2) After >1000 points, depending on learning rate, the predictions overtake your current mouse position and do interesting (still small and with some error) step ahead of your mouse motion.

I still need to check my implementation. I was quite exited to see that the predictions went ahead of my mouse :-).
What I did not think about, is that the mouse speed will play role. So, the problem is somewhat complex.
BZW: What is your network complexity and training rate?

Greetings,
Tolik

**Comment by robertleeplummerjr** [comment]
I updated the demo again based off your recommendations to use at least a 1k inputs: https://jsfiddle.net/robertleeplummerjr/wvkbojr7/3/

> What I did not think about, is that the mouse speed will play role. So, the problem is somewhat complex.

You are right, the problem is a bit complex, and the problem too is that there really isn't any context for the mouse as to what you are moving towards, we are literally just swirling on a canvas, the very basic definition of art...
![image](https://cloud.githubusercontent.com/assets/679099/23129436/afd175b0-f750-11e6-87ff-cb445fb7234f.png)

As for network params:
* 8 inputs neurons, repeating x and y
* 1 hidden layer with 4 neurons
* 2 output neurons
* learning, acceptable error, and training rate I just used defaults found here: https://github.com/harthur-org/brain.js/commit/c86388a7e9652fcf46313931041c454afb0f7714#diff-f0964eec5cfcdf6b9c5b85748849be07R550, which are currently:
```js
NeuralNetwork.trainDefaults = {
  iterations: 20000,
  errorThresh: 0.005,
  log: false,
  logPeriod: 10,
  learningRate: 0.3,
  callback: null,
  callbackPeriod: 10,
  keepNetworkIntact: false
};

NeuralNetwork.defaults = {
  learningRate: 0.3,
  momentum: 0.1,
  binaryThresh: 0.5,
  hiddenLayers: null
};
```

**Comment by robertleeplummerjr** [comment]
I added a generate option here: https://jsfiddle.net/robertleeplummerjr/wvkbojr7/4/
Not very exciting though, it detects no real input, and slows the cursor gradually, and eventually stops, so you end up with a very short line.

**Comment by MariasStory** [comment]
Hi @robertleeplummerjr 
It is not very exiting, but interesting:
![image](https://cloud.githubusercontent.com/assets/6576335/23255876/789f9fb2-f9bd-11e6-80db-024060099c5e.png)
gives:
![image](https://cloud.githubusercontent.com/assets/6576335/23255893/86f87412-f9bd-11e6-8fdd-491d33023d5f.png)
I did try more complex network configurations, but it looks very limited. I guess some optimization is needed.
Gut job. ;-)

**Comment by robertleeplummerjr** [comment]
Nice!  The network is about as simple as they come ( a feedforward neural net, with one hidden layer, and basic backpropagation) so if we gave it more hidden layers, it likely could work out some of the optimization you are talking about.  Too, I think that a feedforward net is going to give us very limited results with the amount of data we are sending it (overfitting, lack of data), a recurrent neural net will probably give us a lot more stable results, but will (as far as I can tell, and I may not be looking at it correctly) will work off the relative position and be more intuitive.

**Comment by robertleeplummerjr** [comment]
I think the only real limitation we have here is overfitting, ie: not enough data.  That being said, if we had a ton of mouse data and resolutions it was tracked with, we could potentially do something very useful with this.  All in all, would you say this is solved?  Also, would you like to see this added to the brainjs.com as a visualization?

**Comment by MariasStory** [comment]
Hi @robertleeplummerjr,
I wanted to see some reasonable predictions of a mouse motion. I did not get it working properly :-(.
Your solution seems to somewhat generalize the training data, as shown on the example.
I don't mind if this example is added to the brainjs.com as a visualization.
Let's close the issue.
Thanks.

**Comment by AmrAlaa-exe** [comment]
@robertleeplummerjr  Hi,
Could you please explain a sample of your code: https://jsfiddle.net/robertleeplummerjr/wvkbojr7/4/ 
using a LSTM? 
Thanks.

**Comment by mubaidr** [comment]
If you still need help, feel free to re-open this issue.
